### PR TITLE
Import/Export/Module

### PR DESCRIPTION
**Pull request by hzoo**
~~Tests are failing so WIP~~

Ref #40

### CODE REVIEW

File: src/elements/types/ExportAllDeclaration.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    }
+
+    _acceptChildren(children) {
+        children.passToken('Keyword', 'export');
+        children.skipNonCode();
+        children.passToken('Punctuator', '*');
+        children.skipNonCode();
+        children.passToken('Identifier', 'from');
+        children.skipNonCode();
+        let source = children.passExpression(); // pass Literal?
```
Comment (hzoo) [inline src/elements/types/ExportAllDeclaration.js]: Should we have a `.passLiteral` or something?

File: src/elements/types/ExportDefaultDeclaration.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    constructor(childNodes) {
+        super('ExportDefaultDeclaration', childNodes);
+    }
+
+    _acceptChildren(children) {
+        children.passToken('Keyword', 'export');
+        children.skipNonCode();
+        children.passToken('Keyword', 'default');
+        children.skipNonCode();
+        let declaration = children.passExpression();  // or declaration
```
Comment (hzoo) [inline src/elements/types/ExportDefaultDeclaration.js]: Similarly, what about `.passDeclaration`?

**Comment by hzoo** [comment]
```js
    passExpressionOrSuper() {
        return this._passExpressionInBraces(expression => expression.isExpression || expression.type === 'Super');
    }
    passExpressionOrSpreadElement() {
        return this._passExpressionInBraces(
            expression => expression.isExpression || expression.type === 'SpreadElement');
    }
```

I think it would be good to do `.passExpression` where you can pass in a function. It sounds better than creating a new function for each case? Or a way for the consumer to pass an array of checks instead of doing ` || expression.type === 'SpreadElement'` in the `elementAssert` file.

Example - I want to check expression or declaration.

This doesn't have to be only for passExpression. Or should I just being doing if statement checks?

File: src/elements/types/ExportAllDeclaration.js (hunk 1 above)
Comment (mdevils) [inline src/elements/types/ExportAllDeclaration.js]: In this case there should be a `children.passNode('Literal')`.

File: src/elements/types/ExportDefaultDeclaration.js (hunk 2 above)
Comment (mdevils) [inline src/elements/types/ExportDefaultDeclaration.js]: I think this is a very custom thing and we don't need to put it to `elementAssert`. Just get `currentElement` and check for `VariableDeclaration` or `FunctionDeclaration` and otherwise call `passExpression()`.

**Comment by mdevils** [comment]
`passExpression` is only useful because it processes situation with braces around. We don't have braces allowed everywhere, so something like `passExpressionOrSpreadElement` would be incorrect as we cannot put braces around `spread` (as far as I understand).

**Comment by hzoo** [comment]
@mdevils Ok I added `passExpressionOrSpreadElement` in a previous commit since I wasn't sure so that will need to be changed. Used in ArrayExpression and CallExpression for spread

**Comment by hzoo** [comment]
Still need to finish `ExportNamedDeclaration` and Imports but otherwise can have a review.

Not sure about the `ModuleDeclaration` and `ModuleSpecifier` stuff

File: src/elements/types/ExportAllDeclaration.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        children.passToken('Keyword', 'export');
+        children.skipNonCode();
+        children.passToken('Punctuator', '*');
+        children.skipNonCode();
+        children.passToken('Identifier', 'from');
+        children.skipNonCode();
+        let source = children.passNode('Literal');
+        children.skipNonCode();
+        children.skipSemicolon();
+        children.assertEnd();
```
Comment (mdevils) [inline src/elements/types/ExportAllDeclaration.js]: Calling `skipNonCode()` and `skipSemicolon()` in a row means we can have trailing spaces.

I would suggest something like this:

```
let source = children.passNode('Literal');
if (!children.isEnd) {
    children.skipNonCode();
    children.passToken('Punctuator', ';');
}
children.assertEnd();
```

Comment (hzoo) [inline src/elements/types/ExportAllDeclaration.js]: Would it be useful it that was a helper?

Comment (hzoo) [inline src/elements/types/ExportAllDeclaration.js]: Ok adding that with a test as well

File: src/elements/types/ExportSpecifier.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+    _acceptChildren(children) {
+        console.log(children._elements.map(function(e) {
+            return e._type;
+        }));
+
+        let local;
+        console.log(children._currentElement);
+        console.log(children.isNode('Identifier'));
+        console.log(children.isToken('Identifier'));
```
Comment (hzoo) [inline src/elements/types/ExportSpecifier.js]: I must be doing something very wrong?

For some reason with `export {x};` `x` is an `Identifier` so `children.passNode('Identifier')` works.
However with  `export {x as y};` `x` is an `Identifier` with  `children.isNode('Identifier')` is false?

Comment (mdevils) [inline src/elements/types/ExportSpecifier.js]: The confusion comes from the fact that there is a node `Identifier` and at the same time there is also a token `Identifier` (which is ofter inside node `Identifier`).

For instance of you have `x = 1`, left side of the expressions is Node `Identifier` with Token `Identifier` inside. But if you have `super()`, then you have a Node `Super` with Token `Identifier` inside.

Comment (hzoo) [inline src/elements/types/ExportSpecifier.js]: Oh ok that makes more sense

Comment (mdevils) [inline src/elements/types/ExportSpecifier.js]: Well, `import {x}` is a shortcut for `import {x as x}`, so if I understand correctly we have:

`import {x}` -> `Token('Keyword', 'import') Token('Punctuator', '{') Identifier([Token('Identifier', 'x')]) Token('Punctuator', '}')`
`import {x as y}` -> `Token('Keyword', 'import') Token('Punctuator', '{') Token('Identifier', 'x') Token('Identifier', 'as') Identifier([Token('Identifier', 'y')]) Token('Punctuator', '}')`

So depending on the input you made decision: whether this is shortcut, or naming an export or even a default (`import {default as z}`).

**Comment by hzoo** [comment]
Ok updated let me know what you think - updated `ExportNamedDeclaration` as well but could probably add more tests with whitespace etc

**Comment by hzoo** [comment]
I guess can finish up this PR with the Import ones.

**Comment by hzoo** [comment]
Ok added imports.

**Comment by hzoo** [comment]
Also we should rename braces -> parens later. `_passExpressionInBraces` -> `_passExpressionInParens` since braces usually refer to `{` and parens to `(`

Not sure https://en.wikipedia.org/wiki/Bracket#Names_for_various_bracket_symbols

**Comment by hzoo** [comment]
How can I do `args.push(children.passExpressionOrSpreadElement());` so that it will pass either an expression or a spread element?

**Comment by hzoo** [comment]
Ready for more review other than points mentioned

**Comment by mdevils** [comment]
> Also we should rename braces -> parens later. _passExpressionInBraces -> _passExpressionInParens since braces usually refer to { and parens to (

I agree, sorry for confusion

**Comment by hzoo** [comment]
I'l do a PR now since it doesn't affect this one

File: src/elements/types/ExportDefaultDeclaration.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+        let declaration;
+        if (children.currentElement.type === 'FunctionDeclaration' ||
+            children.currentElement.type === 'ClassDeclaration') {
+            declaration = children.passStatement();
+        } else {
+            declaration = children.passExpression();
+        }
+
+        children.skipNonCode();
```
Comment (mdevils) [inline src/elements/types/ExportDefaultDeclaration.js]: Same problem with `isEnd` here.

File: src/elements/types/ExportSpecifier.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+            }
+        }
+
+        children.skipNonCode();
+
+        if (!children.isEnd) {
+            children.passToken('Identifier', 'as');
+            children.skipNonCode();
+            exported = children.passNode('Identifier');
+            children.skipNonCode();
```
Comment (mdevils) [inline src/elements/types/ExportSpecifier.js]: Trailing whitespace is possible here

File: src/elements/types/ImportNamespaceSpecifier.js [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+        super('ImportNamespaceSpecifier', childNodes);
+    }
+
+    _acceptChildren(children) {
+        children.passToken('Punctuator', '*');
+        children.skipNonCode();
+        children.passToken('Identifier', 'as');
+        children.skipNonCode();
+        let local = children.passNode('Identifier');
+        children.skipNonCode();
```
Comment (mdevils) [inline src/elements/types/ImportNamespaceSpecifier.js]: Trailing whitespace issue

File: src/elements/types/ImportDefaultSpecifier.js [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+export default class ImportDefaultSpecifier extends ModuleSpecifier {
+    constructor(childNodes) {
+        super('ImportDefaultSpecifier', childNodes);
+    }
+
+    _acceptChildren(children) {
+        children.skipNonCode();
+        let local = children.passNode('Identifier');
+        children.skipNonCode();
```
Comment (mdevils) [inline src/elements/types/ImportDefaultSpecifier.js]: Trailing whitespace issue.

**Comment by mdevils** [comment]
Great job, @hzoo!

I've finished the review. Left several notes about trailing whitespace. The rest is nice.

**Comment by hzoo** [comment]
Ok made changes. Yeah I see that specifiers doesn't need whitespace - covered in declaration already

File: src/elementTree.js [hunk 9]
```diff
@@ -41,17 +41,18 @@ const visitorKeys = {
...
+    /*IT*/ ImportSpecifier: ['imported', 'local'],
     /*IT*/ LabeledStatement: ['label', 'body'],
     /*IT*/ Literal: [],
     /*IT*/ LogicalExpression: ['left', 'right'],
     /*IT*/ MemberExpression: ['object', 'property'],
     /*IT*/ MetaProperty: ['meta', 'property'],
     /*IT*/ MethodDefinition: ['key', 'value'],
-    /*--*/ ModuleSpecifier: [],
+    /*IT*/ ModuleDeclaration: [],
+    /*IT*/ ModuleSpecifier: ['local'],
```
Comment (hzoo) [inline src/elementTree.js]: I also added `local`

File: README.md [hunk 10]
```diff
@@ -54,7 +54,7 @@ Code-related properties:
 * `sourceCode`: generates and returns JavaScript code of the specified `Element`
 * `sourceCodeLength`: returns JavaScript code length
 * `isToken`, `isNode`, `isExpression`, `isStatement`, `isWhitespace`, `isComment`, `isPattern`, `isAssignable`,
-  `isFragment`: code entity flags.
+  `isFragment`, `isModuleDeclaration`, `isModuleSpecifier` : code entity flags.
```
Comment (hzoo) [inline README.md]: Should this be updated too?

Comment (mdevils) [inline README.md]: Yes, this is correct.

**Comment by mdevils** [comment]
Great job! We have almost finished everything for ES6 and some ES7

**Comment by hzoo** [comment]
Maybe make a 0.0.9 release if you want?
